designs/2023-rule-performance-statistics/README.md

```diff
@@ -0,0 +1,328 @@
+
+An analysis of rule performance in ESLint can already be carried out by setting the [TIMING](https://eslint.org/docs/latest/extend/custom-rules#profile-rule-performance) environment variable. However, **additional scripting is still required** to collect/extract more *granular* times data (lint time per file per rule) by following one of the two approaches:
+
+1. Running ESLint per rule, per file and then collecting the file/rule time data output:
+
+    ```bash
+    TIMING=1 DEBUG=eslint:cli-engine eslint --no-eslintrc --rule ... file
+    ```
+
+2. Doing a single ESLint run on all files and then extracting the file/rule time data output:
+
+    ```bash
+    TIMING=all DEBUG=eslint:cli-engine eslint ... files
+    ```
+
+In addition, one needs to create an overview for an effective presentation of the results.
+
+Since ESLint already collects most of this data internally, it would be more *convenient and shareable* to have:
+
+   1. The **times data** from ESLint exposed to the formatters.
+   2. The possibility to also gain more insight into [other runtime statistics](https://github.com/eslint/eslint/issues/14597#issuecomment-1003863524) (not time related).
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+------
+
+1. **Adding the new `stats` option to ESLint:**
+
+ - [_docs/src/use/command-line-interface.md_](https://github.com/mnkiefer/eslint/pull/1/files#diff-80937556f9a8e68352718255c20699fce061fa72760504c6accc3fc6b3b0613aR117): First, we document its *name* and *purpose* for the [Command Line Interface Reference](https://eslint.org/docs/latest/use/command-line-interface) under the *Miscellaneous* category.
+     ```md
+     --stats     Add additional statistics to lint results
+     ```
+
+ - [_lib/options.js_](https://github.com/mnkiefer/eslint/pull/1/files#diff-358c9491edc00f0db6f2f3c317df9aa932135803481b86c9289bd56bf8af0622R375-R3809): Now that the option has been documented, we add the `stats` option to ESLint's CLI options of type (`ParsedCLIOptions`):
+     ```js
+     {
+         option: "stats",
+         type: "Boolean",
+         default: "false",
+         description: "Add statistics to the lint report"
+     }
+     ```
+
+ - [_lib/cli.js_](https://github.com/mnkiefer/eslint/pull/1/files#diff-347ff93ed2b00c93c817863e32fbac5b4fac71d7339a48378980e682777689f4R96): The function `translateOptions(ParsedCLIOptions)` is called, so we add the `stats` option as an input parameter here too.
+
+ - [_lib/eslint/eslint.js_](): The function `processOptions(ESLintOptions)` is called, which is responsible for validating and normalizing options for the *ESLint* CLIEngine instance. Here, we add the `stats` option as a parameter to:
```

Comment:
If we're going to implement this feature only in the flat config mode, then there's no need to change anything in this file since it contains ESLint class for the eslintrc config mode.

Reply:
Ok, I've removed the changes to _lib/eslint/eslint.js_. For _lib/cli.js_ however, the [changes](https://github.com/mnkiefer/eslint/pull/1/files#diff-347ff93ed2b00c93c817863e32fbac5b4fac71d7339a48378980e682777689f4) are still valid, right?

Reply:
> For _lib/cli.js_ however, the [changes](https://github.com/mnkiefer/eslint/pull/1/files#diff-347ff93ed2b00c93c817863e32fbac5b4fac71d7339a48378980e682777689f4) are still valid, right?

Yes, changes in `lib/cli.js` look good to me.

In `lib/options.js`, `stats` should be defined only for the flat config mode, so that the eslint cli reports `--stats` as an unknown option when used in the eslintrc mode, instead of just silently ignoring it. You can see https://github.com/eslint/eslint/pull/17569 for an example of how to conditionally define a CLI option based on the mode (like `--stats`, `--warn-ignored` will be available only in the flat config mode).

Reply:
Ok, I've updated the `stats` option accordingly.